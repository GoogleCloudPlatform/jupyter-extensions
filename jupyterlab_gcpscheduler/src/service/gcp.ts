/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @typescript-eslint/camelcase */
import { ServerConnection } from '@jupyterlab/services';

import {
  CLOUD_FUNCTION_NAME,
  CLOUD_FUNCTION_REGION,
  IMPORT_DIRECTORY,
} from '../data';
import { ProjectStateService } from './project_state';
import {
  handleApiError,
  TransportService,
  GET,
  POST,
} from 'gcp_jupyterlab_shared';
import { IDocumentManager } from '@jupyterlab/docmanager';
import { Notebook } from '@jupyterlab/notebook';

const SERVICE_ACCOUNT_DOMAIN = 'appspot.gserviceaccount.com';
const IMMEDIATE_JOB_INDICATOR = 'jupyterlab_immediate_notebook';
const SCHEDULED_JOB_INDICATOR = 'jupyterlab_scheduled_notebook';

/**
 * Cloud Scheduler Job
 * https://cloud.google.com/scheduler/docs/reference/rest/v1/projects.locations.jobs#Job
 */
export interface CloudSchedulerJob {
  name: string;
  description: string;
  schedule: string;
  timeZone: string;
  httpTarget: {
    body: string;
    headers: { [name: string]: string };
    httpMethod: string;
    uri: string;
    oidcToken: { serviceAccountEmail: string };
  };
}

/** Message type describing an AI Platform training Job */
export interface RunNotebookRequest {
  imageUri: string;
  inputNotebookGcsPath: string;
  jobId: string;
  masterType: string;
  outputNotebookGcsPath: string;
  scaleTier: string;
  region: string;
  acceleratorType: string;
  acceleratorCount: string;
}

/** List of Jobs returned from AI Platform. */
export type ListAiPlatformJobsResponse = gapi.client.ml.GoogleCloudMlV1__ListJobsResponse;

/** AI Platform Job. */
export type AiPlatformJob = gapi.client.ml.GoogleCloudMlV1__Job;

type StorageObject = gapi.client.storage.Object;

const AI_PLATFORM = 'https://ml.googleapis.com/v1';
const CLOUD_SCHEDULER = 'https://cloudscheduler.googleapis.com/v1';
const STORAGE_DOWNLOAD = 'https://storage.googleapis.com/storage/v1';
const STORAGE_UPLOAD = 'https://storage.googleapis.com/upload/storage/v1';
/**
 * Class to interact with GCP services.
 */
export class GcpService {
  private readonly serverSettings = ServerConnection.defaultSettings;
  private readonly runtimeUrl = `${this.serverSettings.baseUrl}gcp/v1/runtime`;

  constructor(
    private _transportService: TransportService,
    private _projectStateService: ProjectStateService,
    private _documentManager: IDocumentManager
  ) {}

  get projectId() {
    return this._projectStateService.projectId;
  }

  set transportService(transportService: TransportService) {
    this._transportService = transportService;
  }

  /**
   * Uploads the specified Notebook JSON representation to the specified path.
   */
  async uploadNotebook(
    notebookContents: string,
    gcsPath: string
  ): Promise<StorageObject> {
    try {
      const { bucket, object } = this._getGcsPathParts(gcsPath);
      const response = await this._transportService.submit<StorageObject>({
        path: `${STORAGE_UPLOAD}/b/${bucket}/o`,
        method: POST,
        headers: { 'Content-Type': 'application/json' },
        body: notebookContents,
        params: {
          name: object,
          uploadType: 'media',
        },
      });
      return response.result;
    } catch (err) {
      console.error(`Unable to upload Notebook contents to ${gcsPath}`);
      handleApiError(err);
    }
  }

  /**
   * Downloads the notebook located at the gcsPath provided and retrieves the file name from the
   * JSON. Creates a new import directory and attempts to rename it, if it fails then creates
   * the notebook in the already-existing directory and deletes the new directory.
   *
   * @param gcsPath Path to the notebook we want to create.
   */
  async importNotebook(gcsPath: string): Promise<Notebook> {
    const contents = await this.downloadNotebook(gcsPath);

    const retrievedName = JSON.parse(contents)
      .metadata?.papermill?.input_path?.split('/')
      .slice(-1)[0];
    const fileName = retrievedName ? retrievedName : 'untitled.ipynb';

    const newDirectory = await this._documentManager.newUntitled({
      type: 'directory',
    });

    try {
      await this._documentManager.rename(newDirectory.path, IMPORT_DIRECTORY);
    } catch (err) {
      // Expect a 409 error if directory exists. Otherwise throw the error
      if (err.message.indexOf('409') === -1) {
        throw err;
      }
      this._documentManager.deleteFile(newDirectory.path);
    }

    return this.createNotebook(fileName, contents);
  }

  private createNotebook(fileName: string, contents: string): Notebook {
    const widget = this._documentManager.createNew(
      IMPORT_DIRECTORY + fileName,
      'Notebook'
    ) as Notebook;
    widget.model.fromString(contents);
    return widget;
  }

  async downloadNotebook(gcsPath: string): Promise<string> {
    try {
      const { bucket, object } = this._getGcsPathParts(gcsPath);
      const response = await this._transportService.submit<string>({
        path: `${STORAGE_DOWNLOAD}/b/${bucket}/o/${encodeURIComponent(object)}`,
        params: {
          alt: 'media',
        },
      });
      return typeof response.result === 'string'
        ? response.result
        : JSON.stringify(response.result);
    } catch (err) {
      console.error(`Unable to download Notebook contents at ${gcsPath}`);
      handleApiError(err);
    }
  }

  /**
   * Submits a Notebook for recurring scheduled execution on AI Platform via a
   * new Cloud Scheduler job.
   * @param request
   * @param regionName
   * @param serviceAccountEmail
   * @param schedule
   */
  async scheduleNotebook(
    request: RunNotebookRequest,
    regionName: string,
    schedule: string
  ): Promise<CloudSchedulerJob> {
    let timeZone = 'America/New_York';
    try {
      timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone;
    } catch (err) {
      console.warn('Unable to determine timezone');
    }

    try {
      const projectId = await this.projectId;
      const locationPrefix = `projects/${projectId}/locations/${regionName}/jobs`;
      const requestBody: CloudSchedulerJob = {
        name: `${locationPrefix}/${request.jobId}`,
        description: SCHEDULED_JOB_INDICATOR,
        schedule,
        timeZone,
        httpTarget: {
          body: btoa(
            JSON.stringify(this._buildAiPlatformJobRequest(request, true))
          ),
          headers: { 'Content-Type': 'application/json' },
          httpMethod: POST,
          oidcToken: {
            serviceAccountEmail: `${projectId}@${SERVICE_ACCOUNT_DOMAIN}`,
          },
          uri: `https://${CLOUD_FUNCTION_REGION}-${projectId}.cloudfunctions.net/${CLOUD_FUNCTION_NAME}`,
        },
      };
      const response = await this._transportService.submit<CloudSchedulerJob>({
        path: `${CLOUD_SCHEDULER}/${locationPrefix}`,
        method: POST,
        body: requestBody,
      });
      return response.result;
    } catch (err) {
      console.error('Unable to create Cloud Scheduler job');
      handleApiError(err);
    }
  }

  /**
   * Submits a Notebook for immediate execution on AI Platform.
   * @param cloudFunctionUrl
   * @param request
   */
  async runNotebook(request: RunNotebookRequest): Promise<AiPlatformJob> {
    try {
      const projectId = await this.projectId;
      const response = await this._transportService.submit<AiPlatformJob>({
        path: `${AI_PLATFORM}/projects/${projectId}/jobs`,
        method: POST,
        body: this._buildAiPlatformJobRequest(request),
      });
      return response.result;
    } catch (err) {
      console.error('Unable to submit Notebook to AI Platform');
      handleApiError(err);
    }
  }

  /**
   * Gets list of AiPlatformJobs
   * @param cloudFunctionUrl
   * @param request
   */
  async listNotebookJobs(
    pageToken?: string
  ): Promise<ListAiPlatformJobsResponse> {
    try {
      const filter = [SCHEDULED_JOB_INDICATOR, IMMEDIATE_JOB_INDICATOR]
        .map(v => `labels.job_type=${v}`)
        .join(' OR ');
      const projectId = await this.projectId;
      const params: { [k: string]: string } = { filter };
      if (pageToken) {
        params['pageToken'] = pageToken;
      }
      const response = await this._transportService.submit<
        ListAiPlatformJobsResponse
      >({
        path: `${AI_PLATFORM}/projects/${projectId}/jobs`,
        params,
      });
      return response.result;
    } catch (err) {
      console.error('Unable to list AI Platform Notebook jobs');
      handleApiError(err);
    }
  }

  async getImageUri(): Promise<string> {
    // TODO need to check if image exist
    const imageUriPrefix = 'gcr.io/deeplearning-platform-release/';
    const runtimeEnv = await this._getRuntimeEnv();
    if (!runtimeEnv) return '';

    const lastDotIndex = runtimeEnv.lastIndexOf('.');
    return `${imageUriPrefix}${runtimeEnv.substr(0, lastDotIndex)}`;
  }

  private async _getRuntimeEnv(): Promise<string> {
    try {
      const response = await ServerConnection.makeRequest(
        this.runtimeUrl,
        { method: GET },
        this.serverSettings
      );
      return await response.text();
    } catch (err) {
      console.warn('Unable to obtain runtime environment');
      return '';
    }
  }

  private _buildAiPlatformJobRequest(
    request: RunNotebookRequest,
    isScheduled = false
  ): AiPlatformJob {
    return {
      jobId: request.jobId,
      labels: {
        job_type: isScheduled
          ? SCHEDULED_JOB_INDICATOR
          : IMMEDIATE_JOB_INDICATOR,
      },
      trainingInput: {
        args: [
          'nbexecutor',
          '--input-notebook',
          request.inputNotebookGcsPath,
          '--output-notebook',
          request.outputNotebookGcsPath,
        ],
        masterConfig: {
          imageUri: request.imageUri,
          acceleratorConfig: {
            count: request.acceleratorCount || undefined,
            type: request.acceleratorType || undefined,
          },
        },
        masterType: request.masterType || undefined,
        region: request.region,
        scaleTier: request.scaleTier,
      },
    } as AiPlatformJob;
  }

  private _getGcsPathParts(
    gcsPath: string
  ): { bucket: string; object: string } {
    if (gcsPath.startsWith('gs://')) {
      gcsPath = gcsPath.slice(5);
    }
    const pathParts = gcsPath.split('/');
    return {
      bucket: pathParts[0],
      object: pathParts.slice(1).join('/'),
    };
  }
}
