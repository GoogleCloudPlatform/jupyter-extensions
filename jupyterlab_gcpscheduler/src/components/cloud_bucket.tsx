import React from 'react';
import TextField from '@material-ui/core/TextField';
import Autocomplete, {
  createFilterOptions,
} from '@material-ui/lab/Autocomplete';
import { withStyles } from '@material-ui/core';
import {
  INPUT_TEXT_STYLE,
  FORM_LABEL_STYLE,
  ALIGN_HINT,
  FieldError,
} from 'gcp_jupyterlab_shared';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import Snackbar from '@material-ui/core/Snackbar';
import { GcpService } from '../service/gcp';
import { Bucket } from '../interfaces';
import FormHelperText from '@material-ui/core/FormHelperText';
const NO_BUCKET_SELECTION = 'Not selected';

interface Props {
  gcpService: GcpService;
  gcsBucket?: string;
  onGcsBucketChange: (newBucketName: string) => void;
}

export interface BucketOption extends Bucket {
  inputValue?: string;
}

interface State {
  value: BucketOption | null;
  isLoading: boolean;
  validBucketOptions: BucketOption[];
  invalidBucketOptions: BucketOption[];
  error?: string;
  openSnackbar: boolean;
}

const StyledAutoComplete = withStyles({
  option: {
    ...INPUT_TEXT_STYLE,
  },
  input: {
    ...INPUT_TEXT_STYLE,
  },
  inputRoot: {
    ...INPUT_TEXT_STYLE,
  },
  root: {
    marginTop: '8px',
    marginBottom: '16px',
  },
})(Autocomplete);

const filter = createFilterOptions<BucketOption | null>();

export class CloudBucketSelector extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      value: null,
      isLoading: false,
      validBucketOptions: [],
      invalidBucketOptions: [],
      openSnackbar: false,
    };
    this.handleSnackbarClose = this.handleSnackbarClose.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleFilterOptions = this.handleFilterOptions.bind(this);
    this.handleGetOptionLabel = this.handleGetOptionLabel.bind(this);
  }

  async componentDidMount() {
    this._getBuckets(this.props.gcsBucket);
  }

  handleSnackbarClose() {
    this.setState({ openSnackbar: false });
  }

  handleChange(event, newValue: BucketOption | string, reason: string) {
    if (reason === 'create-option' || reason === 'select-option') {
      this.setValue(newValue);
    } else if (reason === 'remove-option' || reason === 'clear') {
      this.setValue(null);
    }
  }

  handleFilterOptions(options: BucketOption[], params) {
    const filtered = filter(options, params);
    // Suggest the creation of a new value
    if (params.inputValue !== '') {
      filtered.push({
        inputValue: params.inputValue,
        name: `Create and select "${params.inputValue}"`,
        accessLevel: 'uniform',
      });
    }
    return filtered;
  }

  handleGetOptionLabel(option: string | BucketOption) {
    // Value selected with enter, right from the input
    if (typeof option === 'string') {
      return option;
    }
    const bucketOption = option as BucketOption;
    // Add "xxx" option created dynamically
    if (bucketOption.inputValue) {
      return bucketOption.inputValue;
    }
    // Regular option
    return bucketOption.name;
  }

  render() {
    return (
      <StyledAutoComplete
        value={this.state.value}
        onChange={this.handleChange}
        filterOptions={this.handleFilterOptions}
        selectOnFocus
        clearOnBlur
        handleHomeEndKeys
        id="cloud-bucket-with-create-option"
        options={this.state.validBucketOptions}
        getOptionLabel={this.handleGetOptionLabel}
        noOptionsText={NO_BUCKET_SELECTION}
        renderOption={option => (option as BucketOption).name}
        freeSolo
        size="small"
        loading={this.state.isLoading}
        renderInput={params => (
          <React.Fragment>
            <TextField
              {...params}
              name="gcsBucket"
              variant="outlined"
              margin="dense"
              placeholder={NO_BUCKET_SELECTION}
              label="Cloud Storage bucket"
              fullWidth={true}
              InputLabelProps={{ shrink: true, style: FORM_LABEL_STYLE }}
              error={this.state.error !== undefined}
            />
            {!this.state.error && (
              <FormHelperText style={ALIGN_HINT}>
                Select an existing bucket or create a new one to store results
              </FormHelperText>
            )}
            <FieldError message={this.state.error} />
            <Snackbar
              open={this.state.openSnackbar}
              onClose={this.handleSnackbarClose}
              autoHideDuration={6000}
              message="Created bucket successfully"
              action={
                <IconButton
                  size="small"
                  aria-label="close"
                  color="inherit"
                  onClick={this.handleSnackbarClose}
                >
                  <CloseIcon fontSize="small" />
                </IconButton>
              }
            />
          </React.Fragment>
        )}
      />
    );
  }

  private async _getBuckets(selectedBucketName?: string) {
    try {
      this.setState({ isLoading: true, error: undefined });
      const buckets = await this.props.gcpService.listBuckets();
      const validBucketOptions = [];
      const invalidBucketOptions = [];
      let value = null;
      for (const bucket of buckets.buckets) {
        if (bucket.accessLevel !== 'uniform') {
          invalidBucketOptions.push(bucket);
          continue;
        }
        validBucketOptions.push(bucket);
        if (bucket.name === selectedBucketName) {
          value = bucket;
        }
      }
      this.setState({
        isLoading: false,
        validBucketOptions,
        invalidBucketOptions,
        value,
      });
      this.props.onGcsBucketChange(value ? value.name : null);
    } catch (err) {
      this.setState({
        isLoading: false,
        error: `Unable to retrieve Buckets: ${err}`,
      });
      this.props.onGcsBucketChange(null);
    }
  }

  private async createAndSelectBucket(bucketName: string) {
    try {
      const fineGrain = this.state.invalidBucketOptions.find(
        bucket => bucket.name === bucketName
      );
      if (!fineGrain) {
        await this.props.gcpService.createUniformAccessBucket(bucketName);
        this.props.onGcsBucketChange(bucketName);
        this._getBuckets(bucketName);
        this.setState({ openSnackbar: true });
        return;
      }
      this.setState({
        value: null,
        error:
          'A bucket with incompatible access controls already exists with that name',
      });
      this.props.onGcsBucketChange(null);
    } catch (err) {
      this.setState({ value: null, error: `${err}` });
      this.props.onGcsBucketChange(null);
    }
  }

  private setValue(newValue: BucketOption | string) {
    this.setState({ error: undefined });
    if (!newValue) {
      this.setState({
        value: null,
        error: 'A bucket is required to store results',
      });
      this.props.onGcsBucketChange(null);
      return;
    } else if (typeof newValue === 'string') {
      const existingBucket = this.state.validBucketOptions.find(
        bucket => bucket.name === newValue
      );
      if (!existingBucket) {
        this.createAndSelectBucket(newValue);
      } else {
        this.setState({ value: existingBucket });
        this.props.onGcsBucketChange(newValue);
      }
    } else {
      const newBucketName = newValue.inputValue;
      if (newBucketName) {
        this.createAndSelectBucket(newBucketName);
      } else {
        this.setState({ value: newValue });
        this.props.onGcsBucketChange(newValue.name);
      }
    }
  }
}
