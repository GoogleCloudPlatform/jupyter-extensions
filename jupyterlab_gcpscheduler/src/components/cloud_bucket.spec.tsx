/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @typescript-eslint/no-unused-vars */
import { shallow } from 'enzyme';
import * as React from 'react';
import { CloudBucketSelector, BucketOption } from './cloud_bucket';
import { GcpService } from '../service/gcp';
import { getBucket } from '../test_helpers';

describe('CloudBucketSelector', () => {
  const mockProjectId = jest.fn();
  const mockCreateBucket = jest.fn();
  const mockListBuckets = jest.fn();
  const mockOnGcsBucketChange = jest.fn();
  const mockGcpService = ({
    get projectId() {
      return mockProjectId();
    },
    createUniformAccessBucket: mockCreateBucket,
    listBuckets: mockListBuckets,
  } as unknown) as GcpService;

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('Renders loading', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [getBucket('bucket1'), getBucket('bucket2')],
    });
    mockListBuckets.mockReturnValue(resolvedBuckets);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    expect(component).toMatchSnapshot();
    expect(mockListBuckets).toHaveBeenCalled();
    expect(mockCreateBucket).not.toHaveBeenCalled();
  });

  it('Renders', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [getBucket('bucket1'), getBucket('bucket2')],
    });
    mockListBuckets.mockReturnValue(resolvedBuckets);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await resolvedBuckets;
    expect(component).toMatchSnapshot();
    expect(mockListBuckets).toHaveBeenCalled();
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith(null);
    expect(mockCreateBucket).not.toHaveBeenCalled();
  });

  it('Tries to render with error', async () => {
    const rejectedList = Promise.reject('Error error error');
    mockListBuckets.mockReturnValue(rejectedList);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        gcsBucket="bucket1"
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await rejectedList.catch(() => null);
    expect(mockListBuckets).toHaveBeenCalledTimes(1);
    expect(mockCreateBucket).not.toHaveBeenCalled();
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith(null);
    expect(component.state('error')).toBe(
      'Unable to retrieve Buckets: Error error error'
    );
  });

  it('Tries to render with value not from list', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [getBucket('bucket1'), getBucket('bucket2')],
    });
    mockListBuckets.mockReturnValue(resolvedBuckets);
    shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        gcsBucket="bucket10"
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await resolvedBuckets;
    expect(mockListBuckets).toHaveBeenCalledTimes(1);
    expect(mockCreateBucket).not.toHaveBeenCalled();
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith(null);
  });

  it('Renders with preselected value', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [getBucket('bucket1'), getBucket('bucket2')],
    });
    mockListBuckets.mockReturnValue(resolvedBuckets);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        gcsBucket="bucket2"
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await resolvedBuckets;
    expect(component).toMatchSnapshot();
    expect(mockListBuckets).toHaveBeenCalled();
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith('bucket2');
    expect(mockCreateBucket).not.toHaveBeenCalled();
  });

  it('Renders with preselected value and changes value', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [getBucket('bucket1'), getBucket('bucket2')],
    });
    mockListBuckets.mockReturnValue(resolvedBuckets);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        gcsBucket="bucket2"
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await resolvedBuckets;
    (component.instance() as CloudBucketSelector).handleChange(
      null,
      {
        ...getBucket('bucket1'),
      } as BucketOption,
      'select-option'
    );
    expect(mockListBuckets).toHaveBeenCalled();
    expect(mockCreateBucket).not.toHaveBeenCalled();
    expect(mockOnGcsBucketChange).toHaveBeenCalledTimes(2);
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith('bucket1');
  });

  it('Renders with preselected value and changes value with enter', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [getBucket('bucket1'), getBucket('bucket2')],
    });
    mockListBuckets.mockReturnValue(resolvedBuckets);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        gcsBucket="bucket2"
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await resolvedBuckets;
    (component.instance() as CloudBucketSelector).handleChange(
      null,
      'bucket1',
      'select-option'
    );
    expect(mockListBuckets).toHaveBeenCalled();
    expect(mockCreateBucket).not.toHaveBeenCalled();
    expect(mockOnGcsBucketChange).toHaveBeenCalledTimes(2);
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith('bucket1');
  });

  it('Renders and creates with click', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [getBucket('bucket1'), getBucket('bucket2')],
    });
    const afterCreateResolvedBuckets = Promise.resolve({
      buckets: [
        getBucket('bucket1'),
        getBucket('bucket2'),
        getBucket('new_bucket'),
      ],
    });
    mockListBuckets
      .mockReturnValueOnce(resolvedBuckets)
      .mockReturnValueOnce(afterCreateResolvedBuckets);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await resolvedBuckets;
    (component.instance() as CloudBucketSelector).handleChange(
      null,
      {
        ...getBucket('Create and select "new_bucket"'),
        inputValue: 'new_bucket',
      } as BucketOption,
      'create-option'
    );
    await afterCreateResolvedBuckets;
    expect(mockListBuckets).toHaveBeenCalledTimes(2);
    expect(mockCreateBucket).toHaveBeenCalledWith('new_bucket');
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith('new_bucket');
  });

  it('Renders and creates with enter', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [getBucket('bucket1'), getBucket('bucket2')],
    });
    const afterCreateResolvedBuckets = Promise.resolve({
      buckets: [
        getBucket('bucket1'),
        getBucket('bucket2'),
        getBucket('new_bucket'),
      ],
    });
    mockListBuckets
      .mockReturnValueOnce(resolvedBuckets)
      .mockReturnValueOnce(afterCreateResolvedBuckets);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await resolvedBuckets;
    (component.instance() as CloudBucketSelector).handleChange(
      null,
      'new_bucket',
      'select-option'
    );
    await afterCreateResolvedBuckets;
    expect(mockListBuckets).toHaveBeenCalledTimes(2);
    expect(mockCreateBucket).toHaveBeenCalledWith('new_bucket');
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith('new_bucket');
  });

  it('Renders and tries to create with error', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [getBucket('bucket1'), getBucket('bucket2')],
    });
    const rejectedCreate = Promise.reject('Error error error');
    mockListBuckets.mockReturnValue(resolvedBuckets);
    mockCreateBucket.mockReturnValue(rejectedCreate);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await resolvedBuckets;
    (component.instance() as CloudBucketSelector).handleChange(
      null,
      'new_bucket',
      'select-option'
    );
    await rejectedCreate.catch(() => null);
    expect(mockListBuckets).toHaveBeenCalledTimes(1);
    expect(mockCreateBucket).toHaveBeenCalledWith('new_bucket');
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith(null);
    expect(component.state('error')).toBe('Error error error');
  });

  it('Renders and removes', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [getBucket('bucket1'), getBucket('bucket2')],
    });
    mockListBuckets.mockReturnValue(resolvedBuckets);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        gcsBucket="bucket2"
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await resolvedBuckets;
    (component.instance() as CloudBucketSelector).handleChange(
      null,
      '',
      'remove-option'
    );
    expect(mockListBuckets).toHaveBeenCalled();
    expect(component.state('value')).toBe(null);
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith(null);
    expect(component.state('error')).toBe(
      'A bucket is required to store results'
    );
  });

  it('Renders and tries to create existing fine grain', async () => {
    const resolvedBuckets = Promise.resolve({
      buckets: [
        getBucket('bucket1'),
        getBucket('bucket2'),
        getBucket('bucketfg', false),
      ],
    });
    mockListBuckets.mockReturnValue(resolvedBuckets);
    const component = shallow(
      <CloudBucketSelector
        gcpService={mockGcpService}
        gcsBucket="bucket2"
        onGcsBucketChange={mockOnGcsBucketChange}
      />
    );
    await resolvedBuckets;
    (component.instance() as CloudBucketSelector).handleChange(
      null,
      {
        ...getBucket('Create and select "bucketfg"'),
        inputValue: 'bucketfg',
      } as BucketOption,
      'create-option'
    );
    expect(component).toMatchSnapshot();
    expect(mockListBuckets).toHaveBeenCalled();
    expect(component.state('value')).toBe(null);
    expect(component.state('error')).toBe(
      'A bucket with incompatible access controls already exists with that name'
    );
    expect(mockOnGcsBucketChange).toHaveBeenCalledWith(null);
  });
});
