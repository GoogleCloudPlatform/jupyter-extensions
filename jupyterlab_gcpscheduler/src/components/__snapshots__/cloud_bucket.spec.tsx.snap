// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudBucketSelector Renders 1`] = `
<WithStyles(WithStyles(ForwardRef(Autocomplete)))
  clearOnBlur={true}
  filterOptions={[Function]}
  freeSolo={true}
  getOptionLabel={[Function]}
  handleHomeEndKeys={true}
  id="cloud-bucket-with-create-option"
  loading={false}
  noOptionsText="Not selected"
  onChange={[Function]}
  options={
    Array [
      Object {
        "accessLevel": "uniform",
        "name": "bucket1",
      },
      Object {
        "accessLevel": "uniform",
        "name": "bucket2",
      },
    ]
  }
  renderInput={[Function]}
  renderOption={[Function]}
  selectOnFocus={true}
  size="small"
  value={null}
/>
`;

exports[`CloudBucketSelector Renders and tries to create existing fine grain 1`] = `
<WithStyles(WithStyles(ForwardRef(Autocomplete)))
  clearOnBlur={true}
  filterOptions={[Function]}
  freeSolo={true}
  getOptionLabel={[Function]}
  handleHomeEndKeys={true}
  id="cloud-bucket-with-create-option"
  loading={false}
  noOptionsText="Not selected"
  onChange={[Function]}
  options={
    Array [
      Object {
        "accessLevel": "uniform",
        "name": "bucket1",
      },
      Object {
        "accessLevel": "uniform",
        "name": "bucket2",
      },
    ]
  }
  renderInput={[Function]}
  renderOption={[Function]}
  selectOnFocus={true}
  size="small"
  value={null}
/>
`;

exports[`CloudBucketSelector Renders loading 1`] = `
<WithStyles(WithStyles(ForwardRef(Autocomplete)))
  clearOnBlur={true}
  filterOptions={[Function]}
  freeSolo={true}
  getOptionLabel={[Function]}
  handleHomeEndKeys={true}
  id="cloud-bucket-with-create-option"
  loading={true}
  noOptionsText="Not selected"
  onChange={[Function]}
  options={Array []}
  renderInput={[Function]}
  renderOption={[Function]}
  selectOnFocus={true}
  size="small"
  value={null}
/>
`;

exports[`CloudBucketSelector Renders with preselected value 1`] = `
<WithStyles(WithStyles(ForwardRef(Autocomplete)))
  clearOnBlur={true}
  filterOptions={[Function]}
  freeSolo={true}
  getOptionLabel={[Function]}
  handleHomeEndKeys={true}
  id="cloud-bucket-with-create-option"
  loading={false}
  noOptionsText="Not selected"
  onChange={[Function]}
  options={
    Array [
      Object {
        "accessLevel": "uniform",
        "name": "bucket1",
      },
      Object {
        "accessLevel": "uniform",
        "name": "bucket2",
      },
    ]
  }
  renderInput={[Function]}
  renderOption={[Function]}
  selectOnFocus={true}
  size="small"
  value={
    Object {
      "accessLevel": "uniform",
      "name": "bucket2",
    }
  }
/>
`;
